packages/@lwc/engine-core/src/framework/vm.ts

```diff
@@ -856,3 +856,68 @@ export function forceRehydration(vm: VM) {
         scheduleRehydration(vm);
     }
 }
+
+export function runFormAssociatedCustomElementCallback(vm: VM, faceCb: () => void) {
+    if (!process.env.IS_BROWSER) {
+        return;
+    }
+
+    const {
+        renderMode,
+        shadowMode,
+        def: { formAssociated },
+    } = vm;
+
+    if (shadowMode === ShadowMode.Synthetic && renderMode !== RenderMode.Light) {
+        throw new Error(
+            'Form associated lifecycle methods are not available in synthetic shadow. Please use native shadow or light DOM.'
+        );
```

Comment:
@nolanlawson do you think this should be a dev only error instead?

Reply:
No, I prefer either 1) throwing an error everywhere, or 2) logging a warning in dev mode.